pandas/core/arrays/numpy_.py

```diff
@@ -507,6 +507,23 @@ def to_numpy(
 
         return result
 
+    def _validate_setitem_value(self, value):
+        kind = self.dtype.kind
+        if kind == "b":
+            if lib.is_bool(value):
+                return value
+
+        elif kind == "f":
+            if lib.is_integer(value) or lib.is_float(value):
+                return value
+        else:
+            if lib.is_integer(value) or (lib.is_float(value) and value.is_integer()):
```

Comment:
this is only for integer-dtype.  also you'll need to check itemsize to be sure it fits.  there is probably something in core.dtypes.cast (maybe np_can_hold_element?) that can be reused here